colossus/src/main/scala/colossus/service/ResponseAdapter.scala

```diff
@@ -0,0 +1,45 @@
+package colossus.service
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.language.higherKinds
+
+trait ResponseAdapter[C <: CodecDSL, M[_]] {
+
+  def executeAndMap[T](i : C#Input)(f : C#Output => M[T]) = flatMap(execute(i))(f)
+
+  def execute(i : C#Input) : M[C#Output]
+
+  def flatMap[T](t : M[C#Output])(f : C#Output => M[T]) : M[T]
+
+  def success[T](t : T) : M[T]
+
+  def failure[T](ex : Throwable) : M[T]
+}
+
+trait CallbackResponseAdapter[C <: CodecDSL] extends ResponseAdapter[C, Callback] {
+
+  def client : ServiceClient[C#Input, C#Output]
```

Comment:
maybe this should be `ServiceClientLike`, though I realize we don't really have the equivalent for an `AsyncServiceClient`.  I definitely think there is room to unify those with a HKT similar to what is done here, but we can save that for another PR.